brave-kafka/src/main/java/smartthings/brave/kafka/consumers/TracingConsumerInterceptor.java

```diff
@@ -0,0 +1,128 @@
+        .traceIdHigh(envelope.getTraceIdHigh())
+        .traceId(envelope.getTraceId())
+        .spanId(envelope.getSpanId())
+        .sampled(envelope.getSampled())
+        .shared(envelope.getShared());
+      if (envelope.getParentId() != 0) {
+        builder.parentId(envelope.getParentId());
+      }
+      return TraceContextOrSamplingFlags.create(builder);
+    }
+  };
+
+
+  @Override
+  public ConsumerRecords<K, byte[]> onConsume(ConsumerRecords<K, byte[]> records) {
+    Map<TopicPartition, List<ConsumerRecord<K, byte[]>>> unenvelopedRecords = new HashMap<>(records.count());
+    records.forEach(record -> {
+      try {
+        EnvelopeProtos.Envelope envelope = EnvelopeProtos.Envelope.parseFrom(record.value());
+        TraceContextOrSamplingFlags traceContextOrSamplingFlags = extractor.extract(envelope);
+        TraceContext ctx = traceContextOrSamplingFlags.context();
+        (ctx != null ? tracer.joinSpan(ctx) : tracer.newTrace(traceContextOrSamplingFlags.samplingFlags()))
+          .name("Receive " + record.topic() + "-" + record.partition())
+          .annotate(Constants.SERVER_RECV)
+          .remoteEndpoint(kafkaEndpoint)
+          .flush();
+        // Some of these seem wrong (checksum, value size) but we need to set the timestamp using this constructor
+        ConsumerRecord<K, byte[]> unenvelopedRecord = new ConsumerRecord<>(record.topic(), record.partition(), record.offset(),
+          record.timestamp(), record.timestampType(), record.checksum(), record.serializedKeySize(),
+          record.serializedValueSize(), record.key(), envelope.getPayload().toByteArray());
+
+        TopicPartition tp = new TopicPartition(record.topic(), record.partition());
+        if (unenvelopedRecords.containsKey(tp)) {
+          unenvelopedRecords.get(tp).add(unenvelopedRecord);
+        } else {
+          List<ConsumerRecord<K, byte[]>> list = new LinkedList<ConsumerRecord<K, byte[]>>();
+          list.add(unenvelopedRecord);
+          unenvelopedRecords.put(tp, list);
+        }
+      } catch (InvalidProtocolBufferException e) {
+        e.printStackTrace();
+      }
+    });
+    return new ConsumerRecords<>(unenvelopedRecords);
+  }
+
+  @Override
+  public void onCommit(Map<TopicPartition, OffsetAndMetadata> offsets) {
+
+  }
+
+  @Override
+  public void close() {
+
+  }
+
+  @Override
+  public void configure(Map<String, ?> configs) {
+    if (configs.get("brave.tracer") == null
+      || !(configs.get("brave.tracer") instanceof Tracer)) {
+      throw new ConfigException("brave.tracer", configs.get("brave.tracer"), "Must an be instance of brave.Tracer");
+    } else {
+      tracer = (Tracer) configs.get("brave.tracer");
+    }
+
+    String servers = (String) configs.get("bootstrap.servers");
+    List<InetSocketAddress> addresses = ClientUtils.parseAndValidateAddresses(Arrays.asList(servers.split(",")));
+    // TODO: use the actual server instead of always using the first?
```

Comment:
If we can't get the actual server then I don't think we should include the address otherwise the trace information will be misleading.